File: Domain/Loan.cs
﻿using Domain.Enums;
using Microsoft.EntityFrameworkCore;

namespace Domain
{
    public class Loan
    {
        public int Id { get; set; }
        public LoanType LoanType { get; set; }
        public int? AccountId { get; set; }
        [Precision(18, 2)]
        public decimal CurrentAmount { get; set; } = 0;
        [Precision(18, 2)]
        public decimal FullAmount { get; set; }
        public int CounterpartyId { get; set; }
        public DateTime LoanDate { get; set; } = DateTime.UtcNow.Date;
        public DateTime RepaymentDate { get; set; }
        public string Description { get; set; }
        public string UserId { get; set; }
        public LoanStatus LoanStatus { get; set; } = LoanStatus.InProgress;
        public int CurrencyId { get; set; }

        public User User { get; set; }
        public virtual Account Account { get; set; }
        public virtual Counterparty Counterparty { get; set; }
        public virtual ICollection<Payoff> Payoffs { get; set; } = new List<Payoff>();
        public virtual Currency Currency { get; set; }
    }
}

File: client-app/src/app/models/Loan.ts
import dayjs from "dayjs";
import { Payoff } from "./Payoff";
import { LoanStatus } from "./enums/LoanStatus";
import { LoanType } from "./enums/LoanType";
import { Counterparty } from "./Counterparty";
import { Currency } from "./Currency";

export interface Loan {
    id: number;
    loanType: LoanType;
    accountId: number;
    currentAmount: number;
    fullAmount: number;
    counterpartyId: number;
    loanDate: Date;
    repaymentDate: Date;
    description?: string;
    loanStatus: LoanStatus;
    currencyId: number;
    payoffs: Payoff[];
}

export interface LoanCreateValues {
    loanType: LoanType;
    accountId: number | string;
    fullAmount: number | null;
    counterpartyId: number| string;
    loanDate: dayjs.Dayjs | Date;
    repaymentDate: dayjs.Dayjs | Date;
    description?: string;
}

export interface LoanUpdateValues {
    fullAmount: number | null;
    repaymentDate: dayjs.Dayjs | Date;
    description: string;
}

export interface GroupedLoan {
    counterpartyId: number;
    currencyId: number;
    counterparty: Counterparty;
    currency: Currency;
    creditsCurrentAmount: number;
    debtsCurrentAmount: number;
    creditsFullAmount: number;
    debtsFullAmount: number;
    currentAmount: number;
    fullAmount: number;
    nearestRepaymentDate: Date | null;
    loanType: LoanType;
}